File: src/Java/801-900/859.BuddyStrings.java

import java.awt.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;

class Solution {
    public boolean buddyStrings(String A, String B) {
        if (A.length() != B.length())
            return false;
        LinkedList<Character> r = new LinkedList<>();
        HashSet<Character> s = new HashSet<>();
        for (int i = 0;i < A.length();++i){
            s.add(A.charAt(i));
            if (A.charAt(i) != B.charAt(i)){
                r.add(A.charAt(i));
                r.add(B.charAt(i));
            }
        }
        if (r.size() == 0 && s.size() < A.length())
            return true;
        if (r.size() != 4)
            return false;
        return r.get(0) == r.get(3) && r.get(1) == r.get(2); 
    }
}

File: src/CSharp/801-900/859.BuddyStrings.cs

using System;
using System.Linq;
using System.Collections.Generic;

namespace PeefyLeetCode.BuddyStrings
{
public class Solution {
     public bool BuddyStrings(string A, string B){
        if (A.Length != B.Length)
            return false;
        if (A == B){
            for (int i = 0;i < A.Length;++i)
                for (int j = i + 1; j < A.Length; j++){
                    if (A[i] == A[j])
                        return true;
                }
        }
        List<char> r = new List<char>();
        for (int i = 0;i < A.Length;++i){
            if (A[i] != B[i]){
                r.Add(A[i]);
                r.Add(B[i]);
            }
        }
        return r.Count == 4 && r[0] == r[3] && r[1] == r[2]; 
    }
}
}

